src/Microsoft.ML.Core/Data/ICursor.md

```diff
@@ -0,0 +1,173 @@
+`Next`, and `Combine`. See their documentation for specifics, but they all
+have in common that they treat the `UInt128` as some sort of intermediate hash
+state, then return a new hash state based on hashing of a block of additional
+'bits.' (Since the bits hashed may be fixed, depending on the operation, this
+can be very efficient.) The basic assumption underlying all of that collisions
+between two different hash states on the same data, or hashes on the same hash
+state on different data, are unlikely to collide. Note that this is also the
+reason why `UInt128` was introduced; collisions become likely when we have the
+number of elements on the order of the square root of the hash space. The
+square root of `UInt64.MaxValue` is only several billion, a totally reasonable
+number of instances in a dataset, whereas a collision in a 128-bit space is
+less likely.
+
+Let's consider the IDs of a collection of entities, then, to be ideally an
+"acceptable set." An "acceptable set" is one that is not especially or
+perversely likely to contain collisions versus other sets, and also one
+unlikely to result in an especially or perversely likely to collide set of
+IDs, so long as the IDs are done according to the following operations that
+operate on acceptable sets.
+
+1. The simple enumeration of `UInt128` numeric values from any number is an
+   acceptable set. (This covers how most loaders generate IDs. Typically we
+   start from 0, but other choices, like -1, are acceptable.)
+
+2. The subset of any acceptable set is an acceptable set. (For example, all
+   filter transforms that map any input row to 0 or 1 output rows, can just
+   pass through the input cursor's IDs.)
+
+3. Applying `Fork` to every element of an acceptable set exactly once will
+   result in an acceptable set.
+
+4. As a generalization of the above, if for each element of an acceptable set,
+   you built the set comprised of the single application of `Fork` on that ID
+   followed by the set of any number of application of `Next`, the union of
+   all such sets would itself be an acceptable set. (This is useful, for
+   example, for operations that produce multiple items per input item. So, if
+   you produced two entries based on every single input entry, if the input ID
+   were _id_, then, the first could be `Fork` of _id_, and the second could be
+   `Fork` then `Next` of the same _id_.)
+
+5. If you have potentially multiple acceptable sets, while the union of them
+   obviously might not be acceptable, if you were to form a mapping from each
+   set, to a different ID of some other acceptable set (each such ID should be
+   different), and then for each such set/ID pairing, create the set created
+   from `Combine` of the items of that set with that ID, and then union of
+   those sets will be acceptable. (This is useful, for example, if you had
+   something like a join, or a Cartesian product transform, or somesuch.)
```

Comment:
"somesuch" should be "some such"

Reply:
Boy I really liked that word. And now you've shattered my world by saying it's not a word at all.

---
In reply to: [188818640](https://github.com/dotnet/machinelearning/pull/173#discussion_r188818640) [](ancestors = 188818640)

Reply:
You may deshatter with a hyphen if it helps bind together the words upon which your world is built.